pkg/activator/net/throttler.go

```diff
@@ -0,0 +1,251 @@
+package activator
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"go.uber.org/zap"
+	"k8s.io/apimachinery/pkg/types"
+	"knative.dev/serving/pkg/apis/serving/v1beta1"
+	servinglisters "knative.dev/serving/pkg/client/listers/serving/v1alpha1"
+	"knative.dev/serving/pkg/queue"
+)
+
+type podIPTracker struct {
+	dest     string
+	requests int
+}
+
+type revisionThrottler struct {
+	revID                types.NamespacedName
+	containerConcurrency v1beta1.RevisionContainerConcurrencyType
+
+	// This is a breaker for the revision as a whole. try calls first pass through
+	// this breaker and are either called with clusterIPDest or go through selecting
+	// a podIPTracker and are then called.
+	breaker *queue.Breaker
+
+	// If we have a healthy clusterIPDest this is set to nil.
+	podIPTrackers []*podIPTracker
+
+	// If we dont have a healthy clusterIPDest this is set to the default, otherwise
+	// it is the l4dest for this revision's private clusterIP
+	clusterIPDest string
+
+	mux sync.RWMutex
+
+	logger *zap.SugaredLogger
+}
+
+func newRevisionThrottler(revID types.NamespacedName,
+	containerConcurrency v1beta1.RevisionContainerConcurrencyType,
+	breakerParams queue.BreakerParams) *revisionThrottler {
+	return &revisionThrottler{
+		revID:                revID,
+		containerConcurrency: containerConcurrency,
+		breaker:              queue.NewBreaker(breakerParams),
+	}
+}
+
+func (rt *revisionThrottler) try(ctx context.Context, function func(string) error) error {
+	var ret error
+
+	if !rt.breaker.Maybe(ctx, func() {
+		// See if we can get by with only a readlock
```

Comment:
I thought we actually thought of prefering per-ip, no? To avoid bad loadbalancing decisions?

Reply:
I thought the idea was to go podIP only to help with cold-start (remove network propagation delay). We have to support fallback to clusterIP if no podIPs are available in order to support istio, but I could see some argument for making use of our lb here...

how about we keep with this design as it changes less behavior and should fix the main goals were after, then we can figure out if preferring podip always is a good idea.

Reply:
issue/todo -- we can defer to improvements later.